File: kotlin/src/main/kotlin/parser/ObjectParser.kt
package parser

import lexer.Lexer
import lexer.Token

object ObjectParser {
    private enum class State {
        Default, Value, Comma, Colon, Key,
    }

    fun parse(lexer: Lexer): Result<JsonValue.Object> {
        var state = State.Default
        var map = mapOf<String, JsonValue>()
        var key = ""

        while (true) {
            val token = lexer.getNextToken().getOrElse { return Result.failure(it) }
            when (state) {
                State.Default -> {
                    when (token) {
                        Token.RightBracket -> return Result.success(JsonValue.Object(map))
                        is Token.String -> {
                            key = token.value
                            state = State.Key
                        }
                        else -> return Result.failure(InvalidTokenException(token))
                    }
                }
                State.Key -> {
                    when (token) {
                        Token.Colon -> state = State.Colon
                        else -> return Result.failure(InvalidTokenException(token))
                    }
                }
                State.Colon -> {
                    if (key.isEmpty()) return Result.failure(InvalidTokenException(token))

                    val ret = ValueParser.parse(lexer, token).getOrElse { return Result.failure(it) }
                    map = map.plus(key to ret)
                    state = State.Value
                }
                State.Value -> {
                    when (token) {
                        Token.RightBracket -> return Result.success(JsonValue.Object(map))
                        Token.Comma -> state = State.Comma
                        else -> return Result.failure(InvalidTokenException(token))
                    }
                }
                State.Comma -> {
                    when (token) {
                        is Token.String -> {
                            key = token.value
                            state = State.Key
                        }
                        else -> return Result.failure(InvalidTokenException(token))
                    }
                }
            }
        }
    }
}

File: PHP/src/Parser/ObjectParser.php
<?php

declare(strict_types=1);

namespace Shin1x1\ToyJsonParser\Parser;

use Shin1x1\ToyJsonParser\Lexer\Lexer;
use Shin1x1\ToyJsonParser\Lexer\Token\ColonToken;
use Shin1x1\ToyJsonParser\Lexer\Token\CommaToken;
use Shin1x1\ToyJsonParser\Lexer\Token\EofToken;
use Shin1x1\ToyJsonParser\Lexer\Token\RightCurlyBracketToken;
use Shin1x1\ToyJsonParser\Lexer\Token\StringToken;
use Shin1x1\ToyJsonParser\Parser\Exception\ParserException;

final class ObjectParser
{
    private const STATE_START = 'START';
    private const STATE_KEY = 'KEY';
    private const STATE_COLON = 'COLON';
    private const STATE_COMMA = 'COMMA';
    private const STATE_VALUE = 'VALUE';

    /**
     * @see https://github.com/shin1x1/php8-toy-json-parser/blob/master/diagrams/object_parser.png
     * @return array<string, mixed>
     */
    public static function parse(Lexer $lexer): array
    {
        $array = [];
        $key = '';
        $state = self::STATE_START;

        while (true) {
            $token = $lexer->getNextToken();
            if ($token instanceof EofToken) {
                break;
            }

            switch ($state) {
                case self::STATE_START:
                    if ($token instanceof RightCurlyBracketToken) {
                        return $array;
                    }
                    if ($token instanceof StringToken) {
                        $key = $token->getValue();
                        $state = self::STATE_KEY;
                        break;
                    }
                    throw new ParserException(token: $token);
                case self::STATE_KEY:
                    if ($token instanceof ColonToken) {
                        $state = self::STATE_COLON;
                        break;
                    }
                    throw new ParserException(token: $token);
                case self::STATE_COLON:
                    $array[$key] = ValueParser::parse($lexer, $token);
                    $state = self::STATE_VALUE;
                    break;
                case self::STATE_VALUE:
                    if ($token instanceof RightCurlyBracketToken) {
                        return $array;
                    }
                    if ($token instanceof CommaToken) {
                        $state = self::STATE_COMMA;
                        break;
                    }
                    throw new ParserException(token: $token);
                case self::STATE_COMMA:
                    if ($token instanceof StringToken) {
                        $key = $token->getValue();
                        $state = self::STATE_KEY;
                        break;
                    }
                    throw new ParserException(token: $token);
                default:
                    throw new ParserException(token: $token);
            }
        }

        throw new ParserException(message: 'No end of object');
    }
}

